libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFAuthenticationManager.m

```diff
@@ -454,33 +454,41 @@ - (void)logoutUser:(SFUserAccount *)user
         [[SFPushNotificationManager sharedInstance] unregisterSalesforceNotifications:user];
         [userAccountManager deleteAccountForUser:user error:nil];
         [self revokeRefreshToken:user];
+        // NB: There's no real action that can be taken if this login state transition fails.  At any rate,
+        // it's an unlikely scenario.
+        [user transitionToLoginState:SFUserAccountLoginStateNotLoggedIn];
     }else {
         // Otherwise, the current user is being logged out.  Supply the user account to the
         // "Will Logout" notification before the credentials are revoked.  This will ensure
         // that databases and other resources keyed off of the userID can be destroyed/cleaned up.
-        if ([SFPushNotificationManager sharedInstance].deviceSalesforceId) {
-            [[SFPushNotificationManager sharedInstance] unregisterSalesforceNotifications];
-        }
-        [self cancelAuthentication];
-        [self clearAccountState:YES];
-
-        [self willChangeValueForKey:@"haveValidSession"];
-        [userAccountManager deleteAccountForUser:user error:nil];
-        [self revokeRefreshToken:user];
-        userAccountManager.currentUser = nil;
-        [self didChangeValueForKey:@"haveValidSession"];
-
-        NSNotification *logoutNotification = [NSNotification notificationWithName:kSFUserLogoutNotification object:self];
-        [[NSNotificationCenter defaultCenter] postNotification:logoutNotification];
-        [self enumerateDelegates:^(id<SFAuthenticationManagerDelegate> delegate) {
-            if ([delegate respondsToSelector:@selector(authManagerDidLogout:)]) {
-                [delegate authManagerDidLogout:weakSelf];
+        __weak typeof (self) weakSelf = self;
+        [self clearAccountState:YES withCompletion:^{
```

Comment:
@khawkins Logout should be asynchronous ideally. We've had race-conditions reported by one customer using SSO pages in particular. 1) How about firing a specific notification when  cookie cleanup is complete instead?  2) Document the need to invoke removeAllCookies::withCompletion  and have customers invoke it in such circumstances.